python/ray/data/_internal/plan.py

```diff
@@ -399,6 +399,11 @@ def meta_count(self) -> Optional[int]:
         """
         if self._stages_after_snapshot:
             return None
+        elif self._in_blocks is not None and self._snapshot_blocks is None:
+            # If the plan only has input blocks, we execute it, so snapshot has output.
+            # This applies to newly created dataset. For example, initial dataset from
+            # read, and output datasets of Dataset.split().
+            self.execute()
         # Snapshot is now guaranteed to be the output of the final stage or None.
```

Comment:
This comment seems not quite right, if the input blocks is lazy blocklist, after a `self.execute()` (without force read), this snapshot is still a lazy list, which is not a final output.

Reply:
I can remove the comment if it's confusing, but it just meant to say `_snapshot_blocks` is the latest block to get output from I guess.

Reply:
Updated comment.